### PR TITLE
WebAssembly Support

### PR DESCRIPTION
**Pull request by nneonneo**
This PR adds WebAssembly support to Ghidra, based on my plugin here: https://github.com/nneonneo/ghidra-wasm-plugin/

More details on the plugin can be read on the project README: https://github.com/nneonneo/ghidra-wasm-plugin/#readme

This will resolve Issue #2937.

### CODE REVIEW

File: Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+## Reference Instructions
+:ref.null reftype is opc=0xd0; reftype; sp0_64; ctx_is_directive=0 {
+	sp0_64 = 0;
+}
+:ref.is_null is opc=0xd1; sp1_32; sp1_64; ctx_is_directive=0 {
+	res:4 = zext(sp1_64 == 0);
+	sp1_32 = res;
+}
+:ref.func funcidx is opc=0xd2; funcidx; br_target; sp0_64; ctx_is_directive=0 {
+	sp0_64 = &br_target;
```
Comment (mumbel) [inline Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc:266]: at first glance this looks weird, is this just `ctx_br_target`

File: Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+
+:local.set local4 is opc=0x21; local4; sp1_32; ctx_is_directive=0; ctx_is_op64=0 {
+	local4 = sp1_32;
+}
+:local.set local8 is opc=0x21; local8; sp1_64; ctx_is_directive=0; ctx_is_op64=1 {
+	local8 = sp1_64;
+}
+
+:local.tee local4 is opc=0x22; local4; sp1_32; ctx_is_directive=0; ctx_is_op64=0 {
+	local4 = sp1_32;
```
[the remark this answers is not in the document]
Comment (nneonneo) [reply Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc:322]: Yes, this is already handled in Java; the stack pointer is bumped by 1 slot, so the old value is still on the stack.

File: Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+:return is opc=0x0f; ctx_is_directive=0 {
+	nop();
+	popCallOther(&o0);
+	return [LR];
+}
+:call br_target is opc=0x10; ULeb128; br_target; ctx_is_directive=0 {
+	nop();
+	popCallOther(&i0);
+	call br_target;
+	pushCallOther(&o0);
```
[the remark this answers is not in the document]
Comment (nneonneo) [reply Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc:246]: I believe that the semantics of `call` are that execution resumes at the next pcode op. Ghidra doesn't actually check where the `return` goes to.

File: Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+
+:return is opc=0x0f; ctx_is_directive=0 {
+	nop();
+	popCallOther(&o0);
+	return [LR];
+}
+:call br_target is opc=0x10; ULeb128; br_target; ctx_is_directive=0 {
+	nop();
+	popCallOther(&i0);
+	call br_target;
```
[the remark this answers is not in the document]
Comment (nneonneo) [reply Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc:245]: `LR` is a totally artificial register, so setting LR is actually not necessary because Ghidra will analyze `call`s using the appropriate control flow regardless of any LR register. It's set for most processor modules because it's a real register whose value might matter; in Wasm, the LR register is entirely unobservable so setting it doesn't have any use.

I'm happy to be proven wrong though - I imagine that not setting `LR` might break things like inlined functions, but those are broken anyway so I'm not sure it will matter.

File: Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc (hunk 1 above)
Comment (nneonneo) [reply Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc:266]: Good catch - actually, I should have used `ref.func br_target` rather than `ref.func funcidx` in order to create an operand reference to the function. I'll make that change.

File: Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mumbel) [reply Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc:246]: ah, yeah it does look like other processors do this as well, just tried it with a dummy example and the pcode did show up in decompiler and listing window.

though not sure how this plays with emulator since I think its returning to the next asm instruction not the pcode instruction.

File: Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc (hunk 4 above)
[the remark this answers is not in the document]
Comment (mumbel) [reply Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc:245]: emulator usage will need this

**Comment by nneonneo** [comment]
OK, so it's probably time to talk a bit about the emulator usage. First, I'll admit that I have never used the emulator feature - I was barely aware it existed, so this was a bit of a blind spot for me during development. It seems like this module has some issues with the emulator.

@mumbel Could you walk me through using the emulator with this module (i.e. repro steps - including sample files where applicable), and tell me what you expect / what you see?

**Comment by mumbel** [comment]
`Ghidra/Extensions/SleighDevTools/pcodetest` is a processor module junit test suite using the emulator.  They work with classes in `src/test.processors/java/ghidra/test/processors` that extend `ProcessorEmulatorTestAdapter`.  I'm not sure how much it would take to get emcc into the pcode_defs.py (maybe model it off of the LLVM entry).  I had a 32-bit build laying around and just used the generated files there and its build line slightly modified with emcc instead of gcc.  I didn't try the java portion, but I have a script for another module that finds every `*_Main()` and tries to execute them (a module I'm working on doesn't have a new enough gcc to link pcodetest and breaks the test harness, so got around it with this approach).  I didn't mess around with it much, but it would only execute the first `.locals`, then 1 `.local` and then broke.

```python                                   
from ghidra.app.emulator import EmulatorHelper

func_iter = currentProgram.getFunctionManager().getFunctions(True)
emuHelper = EmulatorHelper(currentProgram)
RETURN_ADDR = None

for main_func in func_iter:
    if not main_func.getName().endswith("_Main"):
        continue
    mainFunctionEntry = main_func.getEntryPoint()
    emuHelper.writeRegister("PC", mainFunctionEntry.getUnsignedOffset())
    emuHelper.writeRegister("SP", 0xf0000000) # random, dont know if it matters
    emuHelper.writeRegister("LR", 0x80000ffa) # addr in _main() calling the _Main()
    
    while not monitor.isCancelled():
        execAddr = emuHelper.getExecutionAddress()
        curr_func = getFunctionContaining(execAddr).getName()
        print "%r %r" % (curr_func, execAddr)
        if curr_func == "_breakOnError":
            exit(1)
        if execAddr == toAddr(emuHelper.readRegister("LR")):
            print "found return"
            break
        if not emuHelper.step(monitor):
            print "ERROR: %r" % emuHelper.getLastError()
            break
 ```

**Comment by nneonneo** [comment]
@mumbel I looked into this, and here are my thoughts.

The first problem with the emulator support boils down to the fact that we use non-flowing context extensively for disassembly. As far as I can tell, the Emulator by default clears all non-flowing context to zero after each instruction, which is not what we want. The short solution to this is to reset the context register at every instruction manually, using something like this:

```python
func = currentProgram.listing.getFunctions("export", "main")[0]

emuHelper = ghidra.app.emulator.EmulatorHelper(currentProgram)
emuHelper.writeRegister(emuHelper.getPCRegister(), func.entryPoint.offset)
emulator = emuHelper.emulator

emulator.setExecuteAddress(func.entryPoint.offset)

contextreg = currentProgram.programContext.baseContextRegister

def updateContextRegister():
    inst = currentProgram.listing.getInstructionAt(emulator.executeAddress)
    if inst:
        context = inst.getRegisterValue(contextreg)
        emulator.setContextRegisterValue(context)

while 1:
    updateContextRegister()
    emulator.executeInstruction(True, monitor)
```

That fixes the issue with `.locals` and `.local` not being properly executed, although it means tweaking your emulator code somewhat.

The second problem is the lack of support for the Wasm-specific Pcode ops, like `popCallOther`. In theory this can be solved by implementing a processor-specific `EmulateInstructionStateModifier`, like what tricore does. I explored implementing this, but ran into an annoying roadblock: `EmulateInstructionStateModifier` subclasses get access to the `Emulate` class, but that class is designed as a "pure" SLEIGH-based emulator which deliberately does not have a reference to the `Program` object: it's meant basically to take a chunk of memory, a language spec, and some initial state, and start executing. However, the current design makes extensive use of the `Program` to store context registers and Wasm module state.

So, I'm probably going to have to partially redesign this to support emulation properly. How critical do you think the emulation feature is?

**Comment by mumbel** [comment]
Ah interesting, yeah I guess the module I used that on before successfully didn't have a register context.

Using the standard harness with the test.processors class instead of that script might solve a lot, I didn't try anything besides that python.

I wouldn't say it's critical as some modules just cannot run the pcodetest (processors w/o modern C compiler), but being able to run that is extremely helpful for validation.

I'll try playing with it a little more with your suggestions and see if it gets further or like you're saying might break on a pcodeop.  I don't recall having one before in the modules I've run through that, so not sure how they're bypassed

**Comment by nneonneo** [comment]
@mumbel Function calls and branches are implemented in part using pcodeops, so not having them will limit you to strictly linear control flow structures. That's probably enough to do some limited testing, at least.

The standard harness will probably not work, given that the context register needs to be reset on each and every instruction to work correctly.

**Comment by nneonneo** [comment]
@mumbel well, I decided to try and implement emulation support. It took longer than expected, but it works pretty well!

I wrote a simple program in C:

I was able to use it to run a simple Hello, World example which uses C stdio functions to print a message, ultimately calling `__syscall3` to invoke the `writev` syscall, which I hook in the emulator - see `emulate_writev.py` for an example. There's no need to set the context stuff manually now, since it's all handled inside `WasmEmulateInstructionStateModifier`.

**Comment by nneonneo** [comment]
Alright, I'm done fussing with the module - it supports pretty much all of WebAssembly 1.1 at this point, and emulation is working pretty smoothly. It needs some external testing and a good review now.

File: Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc (hunk 4 above)
[the remark this answers is not in the document]
Comment (nneonneo) [reply Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc:276]: Fixed via extra emulator support code.

**Comment by nneonneo** [comment]
@mumbel have you had a chance to see if pcodetest works with the new changes implemented?

**Comment by mumbel** [comment]
@nneonneo. No, sorry had some travel come up last week, will try soon.

**Comment by mumbel** [comment]
https://github.com/NationalSecurityAgency/ghidra/blob/5a688dab29c1e81bd0d5c3336ed77c6be5587b63/Ghidra/Extensions/SleighDevTools/pcodetest/c_src/misc_BODY.c#L217
[pcode_conversions.txt](https://github.com/NationalSecurityAgency/ghidra/files/8457548/pcode_conversions.txt)

I hit at least one floating point issue, s0q=bff0000000000000 s1q=41f0000000000000 for that `FLOAT_ALMOST_EQUAL`
(and I had to NOP out some other floating point tests just due to some local changes I can't easily undo)


But it looks like everything else passed in my hacked up setup (hopefully the fact it did hit an error means it did correctly assert other tests w/ my script).

**Comment by mumbel** [comment]
@nneonneo   Noticed I had some undefined bytes between functions, had you looked into decoding/using the `body_size` field for functions.  Before `.locals` there is a uleb128 that is the total size of the locals and instructions.

1.  https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#code-section

**Comment by nneonneo** [comment]
I’m not sure I understood the first comment, due to my unfamiliarity with pcodetest. Does this indicate a bug in the module? It’s definitely a goal to make the floating point stuff work accurately, but there are also a handful of FP instructions that are translated to pcodeops that aren’t currently implemented in the emulator. Is there a minimal example of a bit of WASM code that isn’t producing the right answer under emulation?

As for the undefined bytes, this is intentional but probably a bit surprising. Function definitions in DWARF officially reference the locals as the start of the function, as do most debuggers I’ve tested. So the code size LEB128 isn’t considered part of the function body itself. The Wasm parser does handle and parse these in order to obtain the code boundaries. I can put in a quick fix to give these a data type definition in the listing.

**Comment by nneonneo** [comment]
> @nneonneo Noticed I had some undefined bytes between functions, had you looked into decoding/using the `body_size` field for functions. Before `.locals` there is a uleb128 that is the total size of the locals and instructions.
> 
>     1. https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#code-section

This is fixed now, somewhat hackily, by defining a data item specifically for the size field.

**Comment by nneonneo** [comment]
@mumbel @ryanmkurtz ping…any thoughts on the PR? I think I’ve addressed everything including the emulation stuff, so this should be good to ship. Would be nice to get a sense of where this PR is in terms of getting merged.

**Comment by ryanmkurtz** [comment]
Redirecting my ping to @emteere

**Comment by GhidorahRex** [comment]
@nneonneo Hey, I'd like to get this reviewed ASAP. Could you rebase to the latest master and update? Also update the help.html file with the correct stylesheet `<link rel="stylesheet" type="text/css" href="help/shared/DefaultStyle.css">`.

Also, do you have any samples we can test with?

Appreciate the help. The module looks great. Just want to make sure we iron out the kinks before accepting.

**Comment by jobermayr** [comment]
https://github.com/nneonneo/ghidra-wasm-plugin/issues/15

**Comment by nneonneo** [comment]
@GhidorahRex Will do.

For samples, I use the following sources:

- Random Unity WebAssembly games. There are a lot on the internet. https://www.wasm.com.cn/demo/Tanks/Build/tanks.wasm.code.unityweb is a decent demo. These all tend to be quite large binaries since they incorporate much of the Unity engine code. Il2CppDumper/Il2CppInspector and related tools can be used to assist in recovering metadata (function names and signatures, type definitions, strings, etc.)
- `sendcc.wasm` from DefCon 2019 CTF: https://github.com/o-o-overflow/dc2019f-super-smash-ooos/blob/master/service/src/includes/sendcc.wasm. This one is a moderately sized binary compiled from C++.
- The `wasm` files generated by compiling the tests in https://github.com/WebAssembly/spec. These are useful for testing specific features of the plugin. They're fairly comprehensive, and contain both valid and invalid modules for testing.
- Compiling and testing little programs on my own. Until recently I was using WebAssembly Studio; that's now defunct. I now compile my own modules via Clang; another approach is to use an online compiler like https://wasdk.github.io/WasmFiddle/ (disclaimer: not tested)

**Comment by nneonneo** [comment]
OK, rebased on master as of today, built and tested. Should be good to go.

**Comment by nneonneo** [comment]
@GhidorahRex OK, rebased on master as of today, built and tested. Should be good to go.

**Comment by jobermayr** [comment]
@GhidorahRex Is it OK for you to introduce **new** build warnings?

> Task :WebAssembly:sleighCompile
Compiling ./data/languages/WebAssembly.slaspec:
1178 NOP constructors found
Use -n switch to list each individually
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_154'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_162'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_165'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_166'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_175'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_176'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_178'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_179'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_180'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_187'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_194'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_197'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_198'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_207'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_208'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_210'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_211'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_212'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_226'
Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_238'

1 language successfully compiled

**Comment by mumbel** [comment]
@GhidorahRex    I hacked together a pcodetest.  I'm pretty sure it wont just run under the normal unit test setup, but at least a semi-familiar binary to look at. its my own branch so has some tests I'm currently messing with, (rename from txt to wasm).  I have not really tried this setup and module since April 2022 though (see above) 

[wasm-emcc_O0_pcodetest.txt](https://github.com/NationalSecurityAgency/ghidra/files/11083870/wasm-emcc_O0_pcodetest.txt)


basically took gcc line from log.txt from a `build_*`, replace gcc with `emcc` and tweaked until it worked.

**Comment by GhidorahRex** [comment]
> @GhidorahRex Is it OK for you to introduce **new** build warnings?
> 
> > Task :WebAssembly:sleighCompile
> > Compiling ./data/languages/WebAssembly.slaspec:
> > 1178 NOP constructors found
> > Use -n switch to list each individually
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_154'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_162'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_165'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_166'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_175'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_176'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_178'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_179'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_180'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_187'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_194'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_197'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_198'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_207'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_208'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_210'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_211'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_212'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_226'
> > Leb128_opc2_2b.sinc:1: Unreferenced table: 'opc2_238'
> 
> 1 language successfully compiled

Yes, I think that should be fine. I'll double-check to make sure there's no serious issues with it.

@mumbel That's great, I'll try messing around with it

**Comment by nneonneo** [comment]
I wouldn't mind trying to get pcodetest working for Wasm. I'll need to know how to run pcodetest and what bits of code I should look at. In theory it should not be that hard: basically, using `clang --target=wasm32-wasi` with a suitable SDK installed should be enough to get it to compile. It should even be possible to run the binary using any of the WASI runners such as Wasmtime.

I pushed a couple of changes yesterday to sync with the latest changes from the plugin. Those should be it as far as processor module changes.

**Comment by mumbel** [comment]
You create a `ProcessorEmulatorTestAdapter` class (ex:  https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Processors/PowerPC/src/test.processors/java/ghidra/test/processors/PPC_O0_EmulatorTest.java) and it runs unit tests using a binary file to match that class that you put in Ghidra/Processor/<proc>/data/pcodetests/

https://github.com/NationalSecurityAgency/ghidra/tree/master/Ghidra/Extensions/SleighDevTools/pcodetest

Is where you would integrate into building the binary.  


This is mostly recollection of when I looked at this last year:
The real question is ELF vs wasm, symbol names, debug, etc... that the unit tests assume will be there from a regular gcc/clang build.  I don't know if labels propagate to the point that java can hook correctly into the specific unit test.

You are not actually ever running these binaries, they are entirely emulated (or that's the goal at least, some require qemu assistance).  I don't even think, with they way its written, you could run it for a native compile.   This was actually part of the problem of my hack,  `pcode_test``main`  should link against all the other top level main()s.  For `emcc` it noticed this and the binary was an optimized `main(){return;}`

**Comment by nneonneo** [comment]
Thanks @mumbel. I was able to follow the guidelines you gave to implement pcodetest fully for Wasm. Here are the binaries that were generated by `pcodetest/build.py` and LLVM 15:

[WebAssembly_LLVM.zip](https://github.com/NationalSecurityAgency/ghidra/files/11100979/WebAssembly_LLVM.zip)

This was quite useful: pcodetest found an actual bug in my WebAssembly SLEIGH definition, which was fixed, and now all of the tests pass!

**Comment by mumbel** [comment]
Sweet.  That's a really nice milestone for sleigh verification.  Apologies for not pointing you in that direction last year.

**Comment by jobermayr** [comment]
Merge conflict after 4aa54dd

**Comment by nneonneo** [comment]
@jobermayr fixed

**Comment by nneonneo** [comment]
I've now added support for SIMD binaries in pcodetest, similar to the AVX tests. Under -O3, LLVM will generate binaries that have a few SIMD instructions.

SIMD emulation is not currently implemented, nor am I intending to add SIMD emulation support with this PR. But, at least the test framework exists, and correctly flags the unimplemented pcodeops.

[WebAssembly_LLVM_and_SIMD.zip](https://github.com/NationalSecurityAgency/ghidra/files/11134171/WebAssembly_LLVM_and_SIMD.zip)

**Comment by nneonneo** [comment]
Hmm, so the latest changes to uninitialized memory state handling and the emulation framework (4aa54dd1f913fcc57da9dbb0ee01f6421e52ac10, 1531456b80656d59788b396ccf7cdc3894d90f9d) now break all of my pcodetests. Wasm implicitly stashes all function locals and transfers stack variables to locals during function calls; these behaviours are handled via a "shadow stack" in the Wasm emulation helper. Emulation involves performing bulk memory copies of the register file, regardless of whether or not the register values are fully set: registers are 128 bits wide (for SIMD support) and often only 32-64 bits are actually in use.

I tried calling `memState.setInitialized` on the entire region prior to calling `getChunk`, but to no avail; the "Incomplete read" exception is thrown anyway. I think this is because the pcode tests are using `AdaptedMemoryState`, which has a no-op implementation of `setInitialized`.

I'm not sure what to do here. I guess the emulator is currently in the process of being upgraded, in which case I should do nothing and wait for it to stabilize. Ideally there would be an API for performing a memory copy, maybe even preserving the initialized/uninitialized state of the memory. There's no reasonable way that I can implement the emulator to copy the registers one-at-a-time: this would necessitate tracking the initialized state and data types (sizes) for every local, and restoring that information from the stack on function return.

**Comment by nsadeveloper789** [comment]
I'm gonna check out your branch and see if there's an "easy" way to fix this. At the very least, I should be able to add some override points so your tests can revert to the older emulator while we figure it out.

**Comment by nsadeveloper789** [comment]
Here's what I've come up with:

```diff
--- a/Ghidra/Framework/Emulation/src/main/java/ghidra/pcode/exec/BytesPcodeExecutorStateSpace.java
+++ b/Ghidra/Framework/Emulation/src/main/java/ghidra/pcode/exec/BytesPcodeExecutorStateSpace.java
@@ -188,12 +188,18 @@ public class BytesPcodeExecutorStateSpace<B> {
                        }
                }
 
-               Iterator<ULongSpan> it =
-                       uninitialized.complement(ULongSpan.extent(offset, size)).iterator();
-               if (it.hasNext()) {
-                       ULongSpan init = it.next();
-                       if (init.min().longValue() == offset) {
-                               return readBytes(offset, (int) init.length(), reason);
+               /**
+                * The decoder will buffer ahead, so give it as much as we can, but no more than is actually
+                * initialized. If it's a (non-decode) read, give it everything, but invoke the warning.
+                */
+               if (reason == Reason.EXECUTE_DECODE) {
+                       Iterator<ULongSpan> it =
+                               uninitialized.complement(ULongSpan.extent(offset, size)).iterator();
+                       if (it.hasNext()) {
+                               ULongSpan init = it.next();
+                               if (init.min().longValue() == offset) {
+                                       return readBytes(offset, (int) init.length(), reason);
+                               }
                        }
                }
```

The logic that permits an "incomplete read" should really only apply when buffering bytes for the instruction decoder. This fixes all 3 of the 4 test classes on my end. I assumes the 5 remaining failures are all due to the unimplemented pcodeops you've already mentioned?

Let me know if the patch above gets things back to where they were before the emulaotr changes. Then, we'll decide what to do with it.

**Comment by nsadeveloper789** [comment]
Cool. Thanks for letting me know. The fix should now be in master. Up to you if you want to rebase again or just let git sort it out. I'll let @GhidorahRex know we've sorted this part out.

**Comment by nneonneo** [comment]
@GhidorahRex Any update on the PR? I gather it's being reviewed; just wondering if there's an update on how it's going.

**Comment by hadess** [comment]
Does anything special need to be done to build this support into Ghidra? I applied those patches on top of 10.3.2, and WebAssembly doesn't show up as a possible CPU type when importing a Wasm file.

Running an older version of the stand-alone extension worked as far as detecting the CPU type of the data I was feeding it even if I eventually ran into https://github.com/nneonneo/ghidra-wasm-plugin/issues/15

**Comment by hadess** [comment]
> Does anything special need to be done to build this support into Ghidra?

The special thing that needs to be done is not to forget to commit the manifest change that includes the patch. :man_facepalming:

**Comment by hadess** [comment]
> Resolved then? I gave it a whirl and it all worked, but just in case I rebased everything onto the latest master branch.

Yep, it's working, although I'm kind of stuck as I don't know the first thing about WASM, so I don't know where the entry points are. Hopefully we'll see tutorials in the future when the feature is merged!

**Comment by nneonneo** [comment]
Heads up that there will be some changes incoming to support Ghidra 11. Any word on whether the feature can be accepted, or anything that needs to be changed (besides v11 compatibility)?

**Comment by hadess** [comment]
> Heads up that there will be some changes incoming to support Ghidra 11.

Hey @nneonneo, do you have a rebase/update planned soon? I wanted to avoid losing the feature in the Ghidra Flatpak that we're looking to update to 11.0.

**Comment by nneonneo** [comment]
Updated for 11.0.

**Comment by nneonneo** [comment]
Updated for v11.0 (ping @hadess).

**Comment by nneonneo** [comment]
@GhidorahRex ping, I've updated the module for 11.1 (only one trivial change).

**Comment by nneonneo** [comment]
Yeah, that would be a good idea. The only problem is that I cannot find the exact original rationale for adding this check.

The commit message for this change states:

> The zero-size check added to `BytesPcodeExecutorStateSpace#read` fixes a bug which caused an exception when reading 0 bytes (`offset + size - 1` is not a valid calculation in that case).

This doesn't quite answer *what* was causing zero-byte reads to happen. It probably happened while running the pcodetest test suite, although I don't recall if the read was due to my code. In any case, the function isn't documented as requiring size to be positive, so I think this is a reasonable fix to add to base Ghidra.

**Comment by dgutson** [comment]
> Updated for v11.0 (ping @hadess).

Hi, what's the status of this? Is there any contribution needed?
